### ISSUE TITLE
How to emit events from the server to a client when a callback is triggered from a DB

### ISSUE THREAD

**Issue by PiiXiieeS** [miguelgrinberg/Flask-SocketIO#1144]
## **How to emit events from the server to a client when a callback is triggered from a DB?**
I am adding callbacks in the records of a DB that are triggered when data is changed directly in the DB. And I am willing to emit an event with `socketio.emit` from the callback.

My code is structure in
* run.py
* app/__init__.py
* app/views/chart.py
* app/templates/chart.html

#### run.py
```python
from app import app, socketio
if __name__ == '__main__':
    socketio.run(app=app, host="0.0.0.0", port=5000, debug=True)
```

#### app/__init__.py
```python
from flask import Flask
from flask_socketio import SocketIO

app = Flask(__name__)
socketio = SocketIO(app, engineio_logger=True)

from app.views import chart
```

#### app/views/chart.py
```python
from flask import render_template, request
from app import app, socketio
from DB import *

def my_callback(record):
    print("The record's title is now:", record.title, record.consulting)
    # just add fix 50 to test the effect of the callback in the chart
    socketio.emit('updateData', {'consulting': 50})

@app.route('/chart')
def chart():
    client = DBclient('myDB')
    page = request.args.get('page', default=None, type=str)
    collection = client.getCollection(page)
    legend =collection.name
    result = collection.query('year=2019')
    labels = []
    values = []
    for row in result:
        labels.append(row.month)  # labels = ["January", "February", "March", "April", "May", "June", "July", "August"]
        values.append(row.consulting)   # values = [10, 9, 8, 7, 6, 4, 7, 8]
        row.add_callback(my_callback)    # API to add callbacks when a record in the collection changes 
    return render_template('chart.html', title='Dynamic Chart', page=page,
                           values=values, labels=labels, legend=legend)

@socketio.on('connect')
def connect():
    print('User Connected')

@socketio.on('ButtonEvent')
def ButtonEvent(json):
    print("Button pressed from", json['page'])
    socketio.emit('updateData', {'consulting': 50}, broadcast=True)
```

#### app/templates/chart.html
```html
<!DOCTYPE html>
<html>

<head>
  {% if title %}
  <title>{{ title }}</title>
  {% else %}
  <title>Dynamic Charts from DB!</title>
  {% endif %}

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.3/Chart.min.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.2.0/socket.io.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.9.3/Chart.min.js"></script>
</head>

<body>
  <canvas id="myChart" width="600" height="400"></canvas>
  <p id="caption">The page URL is
    <a href={{ page }}>{{ page }}</a>
  </p>
  <button id="mybutton">Push Me!</button>

  <script type="text/javascript">
    Chart.defaults.global.responsive = false;
    var chartData = {
      labels: [{% for item in labels %}"{{item}}", {% endfor %}],
    datasets: [{
      label: '{{ legend }}',
      fill: true,
      lineTension: 0.1,
      backgroundColor: "rgba(75,192,192,0.4)",
      borderColor: "rgba(75,192,192,1)",
      data: [{% for item in values %}{{ item }}, {% endfor %}],
      spanGaps: false
          }]
    }

    var ctx = document.getElementById("myChart").getContext("2d");
    // create the chart using the chart canvas
    var myChart = new Chart(ctx, {
      type: 'line',
      data: chartData,
    });

    $(document).ready(function () {
      var socket = io.connect('http://' + document.domain + ':' + location.port);

      $('#mybutton').on('click', function () {
        socket.emit('ButtonEvent', { 'page': 'chart.html' });
      });

      socket.on('connect', function () {
        console.log('Connected')
      });

      // when the event is received just apply a random calculation to see the chart changing
      socket.on('updateData', function (data) {
        myChart.data.datasets[0].data[4] = data['consulting'] * Math.floor(Math.random() * 1000000) + 100000;
        console.log(data)
        myChart.update();
      });

    });

  </script>

</body>

</html>
```

In the `html` template I have included a button `#mybutton` to test the emit event `ButtonEvent` from the client to the server and an event `udateData` back from the server to the client to update the chart dynamically. All this works perfectly as expected.

On the other hand when I update directly the DB, although the callback is triggered because I can see the logs writing the message *The record's title is now: Oct 19 732761*, but the event is never received in the client.
```
The record's title is now: Oct 19 732761
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
The record's title is now: Oct 19 5732761
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
```

### Server
$ python run.py
Server initialized for eventlet.
Restarting with stat
Server initialized for eventlet.
Debugger is active!
Debugger PIN: 814-690-851
(98977) wsgi starting up on http://0.0.0.0:5000
(98977) accepted ('127.0.0.1', 58484)
(98977) accepted ('127.0.0.1', 58485)
['Jun 19', 'Nov 19', 'May 19', 'Oct 19', 'Sep 19', 'Aug 19', 'Apr 19', 'Jul 19']
[1593032, 7184910, 1582262, 1732761, 1522685, 1658704, 1232992, 1451402]
127.0.0.1 - - [02/Jan/2020 22:48:16] "GET /chart?page=04c3482baa9e41bb80769b76d0d6a34d HTTP/1.1" 200 3150 5.068020
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet OPEN data {'sid': '572d5ac74a8e48fc8d74bbebb7fbabe8', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
usuario Conectado
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 0
127.0.0.1 - - [02/Jan/2020 22:48:16] "GET /socket.io/?EIO=3&transport=polling&t=MzeEmSp HTTP/1.1" 200 349 0.002460
(98977) accepted ('127.0.0.1', 58512)
572d5ac74a8e48fc8d74bbebb7fbabe8: Received request to upgrade to websocket
127.0.0.1 - - [02/Jan/2020 22:48:16] "GET /socket.io/?EIO=3&transport=polling&t=MzeEmU4&sid=572d5ac74a8e48fc8d74bbebb7fbabe8 HTTP/1.1" 200 183 0.053084
572d5ac74a8e48fc8d74bbebb7fbabe8: Upgrade to websocket successful
572d5ac74a8e48fc8d74bbebb7fbabe8: Received packet MESSAGE data 2["ButtonEvent",{"page":"chart.html"}]
Button pressed from chart.html
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
572d5ac74a8e48fc8d74bbebb7fbabe8: Received packet PING data None
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet PONG data None
572d5ac74a8e48fc8d74bbebb7fbabe8: Received packet PING data None
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet PONG data None
The record's title is now: Oct 19 732761
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
The record's title is now: Oct 19 5732761
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
572d5ac74a8e48fc8d74bbebb7fbabe8: Received packet PING data None
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet PONG data None
572d5ac74a8e48fc8d74bbebb7fbabe8: Received packet MESSAGE data 2["ButtonEvent",{"page":"chart.html"}]
Button pressed from chart.html
572d5ac74a8e48fc8d74bbebb7fbabe8: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
572d5ac74a8e48fc8d74bbebb7fbabe8: Client is gone, closing socket
572d5ac74a8e48fc8d74bbebb7fbabe8: Client is gone, closing socket
127.0.0.1 - - [02/Jan/2020 22:50:31] "GET /socket.io/?EIO=3&transport=websocket&sid=572d5ac74a8e48fc8d74bbebb7fbabe8 HTTP/1.1" 200 0 135.010945
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet OPEN data {'sid': 'd5ec88beb7dc44c4b76f692f8b7bde4c', 'upgrades': ['websocket'], 'pingTimeout': 60000, 'pingInterval': 25000}
usuario Conectado
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet MESSAGE data 0
127.0.0.1 - - [02/Jan/2020 22:50:32] "GET /socket.io/?EIO=3&transport=polling&t=MzeFHap HTTP/1.1" 200 349 0.001061
(98977) accepted ('127.0.0.1', 59079)
d5ec88beb7dc44c4b76f692f8b7bde4c: Received request to upgrade to websocket
127.0.0.1 - - [02/Jan/2020 22:50:32] "GET /socket.io/?EIO=3&transport=polling&t=MzeFHb2&sid=d5ec88beb7dc44c4b76f692f8b7bde4c HTTP/1.1" 200 183 0.000201
d5ec88beb7dc44c4b76f692f8b7bde4c: Upgrade to websocket successful
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet PING data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet PONG data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet PING data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet PONG data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet MESSAGE data 2["ButtonEvent",{"page":"chart.html"}]
Button pressed from chart.html
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet MESSAGE data 2["ButtonEvent",{"page":"chart.html"}]
Button pressed from chart.html
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet MESSAGE data 2["ButtonEvent",{"page":"chart.html"}]
Button pressed from chart.html
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet MESSAGE data 2["updateData",{"consulting":50}]
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet PING data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet PONG data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Received packet PING data None
d5ec88beb7dc44c4b76f692f8b7bde4c: Sending packet PONG data None

### Client
Navigated to http://localhost:5000/chart?page=04c3482baa9e41bb80769b76d0d6a34d
index.js:83 socket.io-client:url parse http://localhost:5000 +0ms
index.js:83 socket.io-client new io instance for http://localhost:5000 +0ms
index.js:83 socket.io-client:manager readyState closed +0ms
index.js:83 socket.io-client:manager opening http://localhost:5000 +0ms
index.js:83 engine.io-client:socket creating transport "polling" +0ms
index.js:83 engine.io-client:polling polling +0ms
index.js:83 engine.io-client:polling-xhr xhr poll +0ms
index.js:83 engine.io-client:polling-xhr xhr open GET: http://localhost:5000/socket.io/?EIO=3&transport=polling&t=MzeEmSp +1ms
index.js:83 engine.io-client:polling-xhr xhr data null +1ms
index.js:83 engine.io-client:socket setting transport polling +3ms
index.js:83 socket.io-client:manager connect attempt will timeout after 20000 +5ms
index.js:83 socket.io-client:manager readyState opening +0ms
index.js:83 engine.io-client:polling polling got data ArrayBuffer(119) +73ms
index.js:83 engine.io-client:socket socket receive: type "open", data "{"sid":"572d5ac74a8e48fc8d74bbebb7fbabe8","upgrades":["websocket"],"pingTimeout":60000,"pingInterval":25000}" +74ms
index.js:83 engine.io-client:socket socket open +0ms
index.js:83 socket.io-client:manager open +74ms
index.js:83 socket.io-client:manager cleanup +0ms
index.js:83 socket.io-client:socket transport is open - connecting +0ms
index.js:83 engine.io-client:socket starting upgrade probes +2ms
index.js:83 engine.io-client:socket probing transport "websocket" +0ms
index.js:83 engine.io-client:socket creating transport "websocket" +1ms
index.js:83 engine.io-client:socket socket receive: type "message", data "0" +1ms
index.js:83 socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +0ms
chart?page=04c3482baa9e41bb80769b76d0d6a34d:78 Connected
index.js:83 engine.io-client:polling polling +9ms
index.js:83 engine.io-client:polling-xhr xhr poll +80ms
index.js:83 engine.io-client:polling-xhr xhr open GET: http://localhost:5000/socket.io/?EIO=3&transport=polling&t=MzeEmU4&sid=572d5ac74a8e48fc8d74bbebb7fbabe8 +0ms
index.js:83 engine.io-client:polling-xhr xhr data null +1ms
index.js:83 engine.io-client:socket probe transport "websocket" opened +55ms
index.js:83 engine.io-client:socket probe transport "websocket" pong +7ms
index.js:83 engine.io-client:socket pausing current transport "polling" +1ms
index.js:83 engine.io-client:polling we are currently polling - waiting to pause +61ms
index.js:83 engine.io-client:polling polling got data ArrayBuffer(4) +17ms
index.js:83 engine.io-client:socket socket receive: type "noop", data "undefined" +17ms
index.js:83 engine.io-client:polling pre-pause polling complete +0ms
index.js:83 engine.io-client:polling paused +1ms
index.js:83 engine.io-client:socket changing transport and sending upgrade packet +1ms
index.js:83 engine.io-client:socket setting transport websocket +0ms
index.js:83 engine.io-client:socket clearing existing transport polling +0ms
index.js:83 engine.io-client:polling ignoring poll - transport state "paused" +1ms
XHR finished loading: GET "<URL>".
XHR finished loading: GET "<URL>".
XHR finished loading: GET "<URL>".
XHR finished loading: GET "<URL>".
XHR finished loading: GET "<URL>".
XHR finished loading: GET "<URL>".
index.js:83 socket.io-client:manager writing packet {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} +18s
index.js:83 socket.io-parser encoding packet {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} +18s
index.js:83 socket.io-parser encoded {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} as 2["ButtonEvent",{"page":"chart.html"}] +0ms
index.js:83 engine.io-client:socket flushing 1 packets in socket +18s
index.js:83 engine.io-client:socket socket receive: type "message", data "2["updateData",{"consulting":50}]" +6ms
index.js:83 socket.io-parser decoded 2["updateData",{"consulting":50}] as {"type":2,"nsp":"/","data":["updateData",{"consulting":50}]} +6ms
index.js:83 socket.io-client:socket emitting event ["updateData",{"consulting":50}] +18s
chart?04c3482baa9e41bb80769b76d0d6a34d:83 {consulting: 50}
index.js:83 engine.io-client:socket writing ping packet - expecting pong within 60000ms +7s
index.js:83 engine.io-client:socket flushing 1 packets in socket +1ms
index.js:83 engine.io-client:socket socket receive: type "pong", data "undefined" +1ms
index.js:83 engine.io-client:socket writing ping packet - expecting pong within 60000ms +25s
index.js:83 engine.io-client:socket flushing 1 packets in socket +0ms
index.js:83 engine.io-client:socket socket receive: type "pong", data "undefined" +3ms
index.js:83 engine.io-client:socket writing ping packet - expecting pong within 60000ms +25s
index.js:83 engine.io-client:socket flushing 1 packets in socket +1ms
index.js:83 socket.io-client:manager writing packet {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} +1m
index.js:83 socket.io-parser encoding packet {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} +1m
index.js:83 socket.io-parser encoded {"type":2,"data":["ButtonEvent",{"page":"chart.html"}],"options":{"compress":true},"nsp":"/"} as 2["ButtonEvent",{"page":"chart.html"}] +0ms
index.js:83 engine.io-client:socket flushing 1 packets in socket +6s
index.js:83 engine.io-client:socket socket close with reason: "ping timeout" +54s
index.js:83 socket.io-client:manager onclose +54s
index.js:83 socket.io-client:manager cleanup +0ms
index.js:83 socket.io-client:socket close (ping timeout) +2m
index.js:83 socket.io-client:manager will wait 584ms before reconnect attempt +1ms
index.js:83 socket.io-client:manager attempting reconnect +585ms
index.js:83 socket.io-client:manager readyState closed +0ms
index.js:83 socket.io-client:manager opening http://localhost:5000 +0ms
index.js:83 engine.io-client:socket creating transport "polling" +588ms
index.js:83 engine.io-client:polling polling +2m
index.js:83 engine.io-client:polling-xhr xhr poll +2m
index.js:83 engine.io-client:polling-xhr xhr open GET: http://localhost:5000/socket.io/?EIO=3&transport=polling&t=MzeFHap +1ms
index.js:83 engine.io-client:polling-xhr xhr data null +0ms
index.js:83 engine.io-client:socket setting transport polling +2ms
index.js:83 socket.io-client:manager connect attempt will timeout after 20000 +4ms
index.js:83 engine.io-client:polling polling got data ArrayBuffer(119) +11ms
index.js:83 engine.io-client:socket socket receive: type "open", data "{"sid":"d5ec88beb7dc44c4b76f692f8b7bde4c","upgrades":["websocket"],"pingTimeout":60000,"pingInterval":25000}" +9ms
index.js:83 engine.io-client:socket socket open +0ms
index.js:83 socket.io-client:manager open +9ms
index.js:83 socket.io-client:manager cleanup +0ms
index.js:83 socket.io-client:socket transport is open - connecting +598ms
index.js:83 socket.io-client:manager reconnect success +1ms
index.js:83 engine.io-client:socket starting upgrade probes +2ms
index.js:83 engine.io-client:socket probing transport "websocket" +1ms
index.js:83 engine.io-client:socket creating transport "websocket" +0ms
index.js:83 engine.io-client:socket socket receive: type "message", data "0" +1ms
index.js:83 socket.io-parser decoded 0 as {"type":0,"nsp":"/"} +55s
chart?page=04c3482baa9e41bb80769b76d0d6a34d:78 Connected
index.js:83 engine.io-client:polling polling +5ms
index.js:83 engine.io-client:polling-xhr xhr poll +15ms
index.js:83 engine.io-client:polling-xhr xhr open GET: http://localhost:5000/socket.io/?EIO=3&transport=polling&t=MzeFHb2&sid=d5ec88beb7dc44c4b76f692f8b7bde4c +0ms
index.js:83 engine.io-client:polling-xhr xhr data null +1ms
index.js:83 engine.io-client:socket probe transport "websocket" opened +5ms
index.js:83 engine.io-client:socket probe transport "websocket" pong +2ms
index.js:83 engine.io-client:socket pausing current transport "polling" +0ms
index.js:83 engine.io-client:polling we are currently polling - waiting to pause +6ms
index.js:83 engine.io-client:polling polling got data ArrayBuffer(4) +2ms
index.js:83 engine.io-client:socket socket receive: type "noop", data "undefined" +3ms
index.js:83 engine.io-client:polling pre-pause polling complete +1ms
index.js:83 engine.io-client:polling paused +0ms
index.js:83 engine.io-client:socket changing transport and sending upgrade packet +0ms
index.js:83 engine.io-client:socket setting transport websocket +1ms
index.js:83 engine.io-client:socket clearing existing transport polling +0ms
index.js:83 engine.io-client:polling ignoring poll - transport state "paused" +1ms

**Comment by PiiXiieeS** [comment]
In fact, I spent several days trying to find the root cause of my problem and by chance I just found it based on the issue #349... while I have no clue why this fixes the problem.
2 first lines added to monkey patch the eventlet package.

```python
from eventlet import monkey_patch as monkey_patch
monkey_patch()
from flask import Flask
from flask_socketio import SocketIO

app = Flask(__name__)
socketio = SocketIO(app, engineio_logger=True)

from app.views import chart
```

**Comment by miguelgrinberg** [comment]
The callback mechanism that you are using likely spawns threads. Standard threads are incompatible with eventlet, they need to be monkey patched to work well. See https://eventlet.net/doc/patching.html for more information.

**Comment by PiiXiieeS** [comment]
Thanks @miguelgrinberg !
While the monkey patching works for Python 3.6:
```python
from eventlet import monkey_patch as monkey_patch
monkey_patch()
```
I am getting the following error with the same code for Python 3.7.5:

TypeError
TypeError: wrap_socket() got an unexpected keyword argument '_context'

Traceback (most recent call last)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 2449, in wsgi_app
response = self.handle_exception(e)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 1866, in handle_exception
reraise(exc_type, exc_value, tb)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
raise value
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
response = self.full_dispatch_request()
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
raise value
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
rv = self.dispatch_request()
File "/Users/piixiiees/.local/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
return self.view_functions[rule.endpoint](**req.view_args)
File "/Users/piixiiees/Documents/Coding/dbflask/app/views/chart.py", line 33, in chart
token_v2="...")
. . .
File "/Users/piixiiees/.local/lib/python3.7/site-packages/requests/sessions.py", line 546, in get
return self.request('GET', url, **kwargs)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
resp = self.send(prep, **send_kwargs)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
r = adapter.send(request, **kwargs)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
timeout=timeout
File "/Users/piixiiees/.local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 672, in urlopen
chunked=chunked,
File "/Users/piixiiees/.local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 376, in _make_request
self._validate_conn(conn)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 994, in _validate_conn
conn.connect()
File "/Users/piixiiees/.local/lib/python3.7/site-packages/urllib3/connection.py", line 394, in connect
ssl_context=context,
File "/Users/piixiiees/.local/lib/python3.7/site-packages/urllib3/util/ssl_.py", line 370, in ssl_wrap_socket
return context.wrap_socket(sock, server_hostname=server_hostname)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/eventlet/green/ssl.py", line 438, in wrap_socket
return GreenSSLSocket(sock, *a, _context=self, **kw)
File "/Users/piixiiees/.local/lib/python3.7/site-packages/eventlet/green/ssl.py", line 76, in __new__
*args, **kw
TypeError: wrap_socket() got an unexpected keyword argument '_context'

**Comment by miguelgrinberg** [comment]
@PiiXiieeS this was an eventlet bug that was fixed a while ago, see https://github.com/eventlet/eventlet/pull/531. Are you on the latest eventlet release?

**Comment by PiiXiieeS** [comment]
@miguelgrinberg, I believe so, I am in 0.25.1

**Comment by miguelgrinberg** [comment]
Yeah, it is supposedly fixed in 0.25.0. Unfortunately there isn't much I can do with this issue. You can switch to gevent and see if you have more luck than with eventlet, or else you can downgrade eventlet until you find a release that works.
